### ISSUE TITLE
[libc+] std::basic_string::insert pessimization for custom character types

### ISSUE THREAD

**Issue by treh** [llvm/llvm-project#56031]
std::basic_string::insert unnecessarily always copies into a `__temp` buffer if `charT` is a custom type.

[The implementation](https://github.com/llvm/llvm-project/blob/c36870c8e79c1690432076c69cb98a879555efa8/libcxx/include/string#L2930-L2949) uses [`__string_is_trivial_iterator`](https://github.com/llvm/llvm-project/blob/c36870c8e79c1690432076c69cb98a879555efa8/libcxx/include/string#L604-L613) to determine if it should copy the input range into a `basic_string __temp` buffer, before calling `__insert_from_safe_copy`.

However, for custom class-type `_CharT`, even a plain `_CharT const*` does not satisfy `__string_is_trivial_iterator`, so nothing is won by creating this buffer.

Shouldn't the definition for `__string_is_trivial_iterator` use something like `is_nothrow_convertible<Tp, _CharT>` instead of `is_arithmetic<Tp>`?

Note: Custom class-type character types **are** allowed in `basic_string`, the [requirements are](https://timsong-cpp.github.io/cppwp/n4861/strings#basic.string-1)

> The class template basic_­string describes objects that can store a sequence consisting of a varying number of arbitrary char-like objects with the first element of the sequence at position zero. Such a sequence is also called a “string” if the type of the char-like objects that it holds is clear from context. In the rest of this Clause, the **type of the char-like objects** held in a basic_­string object is designated by **charT**.

and *char-like* is defined [before](https://timsong-cpp.github.io/cppwp/n4861/strings#general-1) as

> This Clause describes components for manipulating sequences of **any non-array trivial standard-layout type. Such types are called char-like types**, and objects of char-like types are called char-like objects or simply characters.

Note2: This even was a bug from [this]( https://reviews.llvm.org/D98573) to [this](https://reviews.llvm.org/D119633) revision, causing infinite recursion.

Here is a small example to play around with: [main.cpp](https://github.com/llvm/llvm-project/files/8899669/main.cpp.txt)

**Comment by philnik777** [comment]
Wait, no it doesn't work. See https://quuxplusone.github.io/blog/2021/04/17/pathological-string-appends/#self-appending-via-conversion-operators for a detailed explanation.

**Comment by frederick-vs-ja** [comment]
`is_nothrow_*` doesn't work because the conversion may do bad things, e.g., unwanted aliasing.

I think the current implementation can be improved:
1. forward iterators must provide multipass gutantee, so it' safe to calculate `distance`. (BTW, some throwing-on-tenth-increment iterator doesn't meet the requirements of forward itetator.)
2. Whenever the final length of string is known, if the final length is greater than the capacity, we should direct construct the new content in the new storage. Otherwise, if SSO

**Comment by frederick-vs-ja** [comment]
No, `is_nothrow_convertible` doesn't work, because the iterator or the conversion from its value type may unexpectedly alias the content in the current string.

It seems that unary `__string_is_trivial_iterator` is insufficient. It may be better to have `__string_is_trivial_iterator<Iter, CharT>` and `__string_is_trivial_iterator<cv T*, T>` should always be `true`.

On the other hand, IMO `__string_is_trivial_iterator<std::list<int>::iterator>::value` should be `true`. Perhaps libc++ should have a opt-in mechanism to tell whether an iterator doesn't involve potentially unwanted aliasing.

----
@philnik777 @Quuxplusone 

However, I do believe we can generally improve insertion of `basic_string` in the cases of forward iterators.

First, I do believe `std::distance` can be called for arbitrary times on a range represented by forward iterator (or `std::ranges::distance` for `std::ranges::forward_range`) due to the multi-pass guarantee ([[forward.iterator]/3](https://eel.is/c++draft/forward.iterators#3), [[iterator.concept.forward]/4](https://eel.is/c++draft/iterator.concept.forward#4)). So I think @Quuxplusone's [`ThowIt`](https://quuxplusone.github.io/blog/2021/04/17/pathological-string-appends/#iterator-operations-that-throw-exceptions) is not a valid forward iterator and it is undefined behavior to pass it to standard library functions.

Second, I think we can and **generally should construct the newly inserted contents before relocating existing contents**.
When the final length of the resulted string can be safely known, I suggest the following logic:

1. If the resulted string doesn't fit in the existing storage, we should **allocate exactly once** for the new storage and then directly construct the new contents at the required location. Relocation should be done after the construction.
2. If the resulted string fits in the existing storage and the SSO buffer is not used, **no allocation should be performed**, and we should construct the new contents after the terminating null character, and then rotate the string. (Or when the rotation may be not cheap, a temporary non-dynamic buffer can be used.)
3. If the resulted string fits in the SSO buffer, then we should created a temporary (short) string before insertion to avoid aliasing. Note that no dynamic allocation is performed in this case.

Note that the while the terminating null character is not an element, it must be treated as an element during insertion, so probably `append` can't be simpler than `insert`.

----
As fas as I known, the `__string_is_trivial_iterator` stuffs are used to detect whether the existing contents can be safely relocated **before** the construction of new contents. I don't think the intent is generally right.

Perhaps the stuffs should only be used in some cases for `replace`:
- there are some new contents to insert and the some old contents to remove, and
- the resulted string fits in the existing allocated storage, and, however,
- **the new contents don't fit in the spare after elements (including the terminating null character)**.

In such a case, if insertion can be done safely after removal of old contents, we can avoid allocation. Otherwise, the allocation seems unavoidable.

Anyway, I agree with @Quuxplusone on the conclusion that `is_nothrow_*` or `noexcept` is unhelpful during insertion.

**Comment by philnik777** [comment]
> On the other hand, IMO `__string_is_trivial_iterator<std::list<int>::iterator>::value` should be `true`. Perhaps libc++ should have a opt-in mechanism to tell whether an iterator doesn't involve potentially unwanted aliasing.

We could add something like `__is_container_iterator`.

> However, I do believe we can generally improve insertion of `basic_string` in the cases of forward iterators.
> 
> First, I do believe `std::distance` can be called for arbitrary times on a range represented by forward iterator (or `std::ranges::distance` for `std::ranges::forward_range`) due to the multi-pass guarantee ([[forward.iterator]/3](https://eel.is/c++draft/forward.iterators#3), [[iterator.concept.forward]/4](https://eel.is/c++draft/iterator.concept.forward#4)). So I think @Quuxplusone's [`ThowIt`](https://quuxplusone.github.io/blog/2021/04/17/pathological-string-appends/#iterator-operations-that-throw-exceptions) is not a valid forward iterator and it is undefined behavior to pass it to standard library functions.

I think you are right. Having `is_same_v<iterator::iterator_category, forward_iterator_tag>` and then throw unpredictably is undefined behaviour.

> Second, I think we can and **generally should construct the newly inserted contents before relocating existing contents**. When the final length of the resulted string can be safely known, I suggest the following logic:
> 
>     1. If the resulted string doesn't fit in the existing storage, we should **allocate exactly once** for the new storage and then directly construct the new contents at the required location. Relocation should be done after the construction.
> 
>     2. If the resulted string fits in the existing storage and the SSO buffer is not used, **no allocation should be performed**, and we should construct the new contents after the terminating null character, and then rotate the string. (Or when the rotation may be not cheap, a temporary non-dynamic buffer can be used.)

I think for `append` we could do
```c++
auto __first_begin = __first;
while (__first != __last) {
  traits_type::assign(*__whatever_pos, *__first);
  ++__first;
  ++__whatever_pos;
}
traits_type::assign(*__whatever_pos_begin, *__first_begin);
```
That would leave the original string unmodified until we re-assign the null-terminator and would still construct all the content in-place.

>     3. If the resulted string fits in the SSO buffer, then we should created a temporary (short) string before insertion to avoid aliasing. Note that no dynamic allocation is performed in this case.
> 
> Note that the while the terminating null character is not an element, it must be treated as an element during insertion, so probably `append` can't be simpler than `insert`.
> 
> As fas as I known, the `__string_is_trivial_iterator` stuffs are used to detect whether the existing contents can be safely relocated **before** the construction of new contents. I don't think the intent is generally right.
> 
> Perhaps the stuffs should only be used in some cases for `replace`:
> 
>     * there are some new contents to insert and the some old contents to remove, and
> 
>     * the resulted string fits in the existing allocated storage, and, however,
> 
>     * **the new contents don't fit in the spare after elements (including the terminating null character)**.
> 
> 
> In such a case, if insertion can be done safely after removal of old contents, we can avoid allocation. Otherwise, the allocation seems unavoidable.
> 
> Anyway, I agree with @Quuxplusone on the conclusion that `is_nothrow_*` or `noexcept` is unhelpful during insertion.

Would you like to make a PR or should I do? I know `string` quite well by now, but it's your idea and I don't want to steal it from you.
